ARCHITECTURE.md

```diff
@@ -0,0 +1,140 @@
+to `ImageStreamTag`s in the `stable` `ImageStream` within the test `Namespace`.
+
+In order to ensure that resources from tests do not leak to the cluster the tests
+are executed on, both hard and soft TTLs are set on the `Namespace` and the
+[`ci-ns-ttl-controller`](https://github.com/openshift/ci-ns-ttl-controller) is
+used to enforce the TTLs and reap namespaces when TTLs have expired.
+
+Namespace initialization follows input resolution in the `ci-operator` output:
+
+```
+2018/10/08 05:28:45 Creating namespace ci-op-31xmgx1s
+2018/10/08 05:28:47 Creating rolebinding for user flaper87 in namespace ci-op-31xmgx1s
+2018/10/08 05:28:47 Setting a soft TTL of 1h0m0s for the namespace
+2018/10/08 05:28:47 Setting a hard TTL of 12h0m0s for the namespace
+2018/10/08 05:28:47 Setting up pipeline imagestream for the test
+2018/10/08 05:28:47 Tagging openshift/release:golang-1.10 into pipeline:root
+2018/10/08 05:28:47 Tagging openshift/origin-v4.0:base into pipeline:base
+2018/10/08 05:28:47 Tagged release images from openshift/origin-v4.0:${component}, images will be pullable from registry.svc.ci.openshift.org/ci-op-31xmgx1s/stable:${component}
+```
+
+## Build Graph Traversal
+
+A configuration file for the CI Operator defines build steps, test targets and
+output images for a component git repository. A graph of build dependencies is
+built from this configuration in order to determine what concrete actions need
+to occur for any specific target. Each invocation of `ci-operator` specifies one
+or more `--target`s to execute; for each target, the build graph is traversed to
+execute dependent steps first.
+
+The CI Operator configuration file creates some implicit build steps:
+
+| Output `ImageStreamTag` | Action |
+| ----------------------- | ------ |
+| `pipeline:src`          | clones the refs under test |
+| `pipeline:bin`          | runs the [`binary_build_commands`](./CONFIGURATION.md#binary_build_commands) |
+| `pipeline:test-bin`     | runs the [`test_binary_build_commands`](./CONFIGURATION.md#test_binary_build_commands) |
+| `pipeline:rpms`         | runs the [`rpm_build_commands`](./CONFIGURATION.md#rpm_build_commands) |
+
+Container image builds -- whether from the implicit `pipeline` steps above or
+from explicit image build configurations in the [`images`](./CONFIGURATION.md#images)
+stanza, are executed using OpenShift `Build`s. Test targets in the [`tests`](./CONFIGURATION.md#tests)
+stanza are executed using Kubernetes `Pod`s. As all of the test workflow execution
+objects are created in a `Namespace` shared for all jobs with the same input,
+re-use is achieved by deterministic naming. For instance, the `src` `Build` that
+creates the `pipeline:src` `ImageStreamTag` will be created only once in a given
+`Namespace`; other builds of jobs that require this build step will see the
+`Build` running and wait for it to complete or see the `ImageStreamTag` existing
+and consider the build step finished.
```

Comment:
One thing missing from all our documentation files that I was hoping to see here is the different types of image builds that are supported and the corresponding `ci-operator` configuration entry. A few examples extracted quickly from the existing configuration files:

```yaml
- dockerfile_path: test/install/Dockerfile.Fedora
  from: bin
  to: fedora-install
- dockerfile_path: images/installer/Dockerfile.ci
  from: base
  inputs:
    root:
      as:
      - build
  to: installer
- from: os
  inputs:
    root:
      as:
      - '0'
  to: release-controller
- dockerfile_path: Dockerfile.e2e
  from: base
  inputs:
    bin:
      paths:
      - destination_dir: .
        source_path: /go/src/github.com/openshift/openshift-azure/e2e.test
    root:
      as:
      - e2e-tests
  to: e2e-tests
```

Reply:
Me too, but I wonder if architecture description is the right place to put these. I start to think we might need a cookbook with examples - not just for images but also for tests, especially now with the new test types.

Reply:
The configuration doc points to the OpenShift Build documents -- this is just using OpenShift Build flows

Reply:
/shrug